### PR TITLE
fix LocalCatalogEntry.get() so that a new object isn't created each time it is accessed

### PR DESCRIPTION
**Pull request by gwbischof**
fixes https://github.com/bluesky/databroker/issues/516

### CODE REVIEW

**Comment by gwbischof** [comment]
What is happening is that when we read partition 0, we call load_partitions because self._partitions is None, and then return partition 0. Then we read_partition 1, self._partitions in None again, so we call load_partitions. We try to return partition 1, but we find that there are missing datum, so we insert the missing datum into self._partitions at index 1. This shifts the old partition 1 to partition 2. Then we return the new partition 1. Then we try to get the next partition (partition 2). self._partitions is None again, so we call load_partitions which gives us partition 0 and partition 1. When we try to return partition 2, we find that there is no partition 2, so we return an empty list. If self._partitions kept its value this would work correctly.

File: databroker/core.py [hunk 1]
```diff
@@ -193,6 +196,19 @@ def __getstate__(self):
                               for k, arg in self._captured_init_kwargs.items()})
         return OrderedDict(cls=self.classname, args=args, kwargs=kwargs)
 
+    def get(self, **user_parameters):
+        if self._data_source is None:
```
Comment (danielballan) [inline databroker/core.py]: The second time `get` is called (and thereafter) the `user_parameters` will be ignored and we will get the result from whatever `user_parameters` were given the first time `get` was called.

The `Entry` class just above attempts to do this caching in a way that accounts for the `user_parameters` a.k.a `kwargs`. Would it work to make `StreamEntry` inherit from `Entry` and reuse its mechanism?

Comment (gwbischof) [reply databroker/core.py]: I dont think the LRU cache will work for this.
I think you want to permanently cache the streams or else you can have the same issue again.

File: databroker/core.py [hunk 2]
```diff
@@ -193,6 +196,21 @@ def __getstate__(self):
...
+            data_source = plugin(**open_args)
+            data_source.catalog_object = self._catalog
+            data_source.name = self.name
+            data_source.description = self._description
+            data_source.cat = self._catalog
+            self._data_source = data_source
+            return data_source
+        else:
+            _, open_args = self._create_open_args(user_parameters)
+            self._data_source.__init__(**open_args)
```
Comment (tacaswell) [inline databroker/core.py]: If we are going to re-run `__init__` why not just start from scratch?

Comment (gwbischof) [reply databroker/core.py]: Updated this, it doesn't work this way any longer.

File: databroker/core.py (hunk 1 above)
Comment (gwbischof) [reply databroker/core.py]: Updated

**Comment by gwbischof** [comment]
I just tested this again at CSX with the latest changes, and it is working correctly.

**Comment by danielballan** [comment]
Rebased and force-pushed to remove presumably accidentally-committed file `bluejeans-v2_2.0.0-189_amd64.deb`.

Notes from call:
* This reuses the same approach that was we use with `Entry` and `BlueskyRun` to avoid instantiating the same thing many times.
* We use a `dict` as the cache instead of `LRUCache` because it is critical that the same instance be reused. It's hard to see how that cache could grow beyond length-1.
* The source of the issue is fragility around how partitions interact with Resources in the case when Resources do not have a `run_start` field. We will revisit this logic once we can assume that _all_ Resources know their RunStart.

**Pull request merged**